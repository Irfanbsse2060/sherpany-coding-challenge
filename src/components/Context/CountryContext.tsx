import React from 'react'

type Action = {type: 'update', payload: string[]}
type Dispatch = (action: Action) => void
type State = {countries: string[]}
type CountryProviderProps = {children: React.ReactNode}


const CountryStateContext = React.createContext<State | undefined>(undefined)
const CountryDispatchContext = React.createContext<Dispatch | undefined>(
    undefined,
)

function countryReducer(state: State, action: Action) {
    switch (action.type) {
        case 'update': {
            return {countries: action.payload}
        }
        default: {
            throw new Error(`Unhandled action type: ${action.type}`)
        }
    }
}



function CountryProvider({children}: CountryProviderProps) {
    const [state, dispatch] = React.useReducer(countryReducer, {countries: []})
    return (
        <CountryStateContext.Provider value={state}>
        <CountryDispatchContext.Provider value={dispatch}>
            {children}
            </CountryDispatchContext.Provider>
            </CountryStateContext.Provider>
    )
}


function useCountryState() {
    const context = React.useContext(CountryStateContext)
    if (context === undefined) {
        throw new Error('useCountState must be used within a CountProvider')
    }
    return context
}
function useCountryDispatch() {
    const context = React.useContext(CountryDispatchContext)
    if (context === undefined) {
        throw new Error('useCountDispatch must be used within a CountProvider')
    }
    return context
}


export {CountryProvider, useCountryState, useCountryDispatch}
