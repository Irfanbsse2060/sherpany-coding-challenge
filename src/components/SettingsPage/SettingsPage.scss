@import "../../styles/variables";

.settings-page {
  display: grid;
  grid-template-columns: 1fr;
  grid-row-gap: 2rem;
  grid-column-gap: 2rem;
  padding: 2rem;
}

