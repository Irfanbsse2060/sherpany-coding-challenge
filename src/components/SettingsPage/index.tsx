import React from 'react';
import {CheckBox} from '../UI'
import {useCountryDispatch, useCountryState} from '../Context/CountryContext'

import './SettingsPage.scss'

function SettingsPage() {
    const dispatch = useCountryDispatch()
    const {countries} = useCountryState()
    const [selectedCountries, setSelectedCountries] = React.useState({
        'CH': countries.includes('CH'),
        'ES': countries.includes('ES'),
        'FR': countries.includes('FR'),
        'GB': countries.includes('GB'),
    })

    const countriesList = [
        {isChecked: selectedCountries['CH'], label: 'Canada (CH)', name: 'CH', value: 'CH'},
        {isChecked: selectedCountries['ES'], label: 'Spain (ES)', name: 'ES', value: 'ES'},
        {isChecked: selectedCountries['FR'], label: 'France (FR)', name: 'FR', value: 'FR'},
        {isChecked: selectedCountries['GB'], label: 'United kingdom (GB)', name: 'GB', value: 'GB'},
    ]

    const handleChange = (name: string, isChecked: boolean) => {
        const updatedSelectedCountriesState = {
            ...selectedCountries,
            [name]: isChecked
        }
        setSelectedCountries(updatedSelectedCountriesState)
        dispatch({
            type: 'update',
            // @ts-ignore
            payload: Object.keys(updatedSelectedCountriesState).filter(country => updatedSelectedCountriesState[country])
        })
    }

    return (
        <main className='settings-page'>
            <h3> Select Nationality to filter user</h3>
            <CheckBox options={countriesList} handleChange={handleChange}/>
        </main>
    );
}

export default SettingsPage;

