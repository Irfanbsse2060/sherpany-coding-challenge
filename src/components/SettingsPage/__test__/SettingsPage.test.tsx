import React from "react";
import {render} from '@testing-library/react';
import SettingsPage from '../index'
import {CountryProvider} from "../../Context/CountryContext";


describe('SettingPage', () => {
    const renderComponent = () => {
        return (
            <CountryProvider>
                <SettingsPage/>
            </CountryProvider>
        )
    }

    test('should be able to render the component', () => {
        const {container} = render(renderComponent());
        expect(container).toMatchSnapshot()
    });

    test('should be able to select country', () => {
        const {queryByText} = render(renderComponent());
        const selector = '[type="checkbox"][value="CH"]'

        // @ts-ignore
        expect(document.querySelector(selector).checked).not.toBeTruthy()
        // @ts-ignore
        document.querySelector(selector).click()
        // @ts-ignore
        expect(document.querySelector(selector).checked).toBeTruthy()
    });


})
