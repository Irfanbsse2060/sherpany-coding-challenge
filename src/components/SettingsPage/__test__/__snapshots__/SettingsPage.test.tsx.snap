// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingPage should be able to render the component 1`] = `
<div>
  <main
    class="settings-page"
  >
    <h3>
       Select Nationality to filter user
    </h3>
    <ul
      class="check-box"
      data-testid="check-box"
    >
      <li
        class="check-box__item"
      >
        <input
          id="CH"
          name="CH"
          type="checkbox"
          value="CH"
        />
        <label
          for="CH"
        >
          Canada (CH)
        </label>
      </li>
      <li
        class="check-box__item"
      >
        <input
          id="ES"
          name="ES"
          type="checkbox"
          value="ES"
        />
        <label
          for="ES"
        >
          Spain (ES)
        </label>
      </li>
      <li
        class="check-box__item"
      >
        <input
          id="FR"
          name="FR"
          type="checkbox"
          value="FR"
        />
        <label
          for="FR"
        >
          France (FR)
        </label>
      </li>
      <li
        class="check-box__item"
      >
        <input
          id="GB"
          name="GB"
          type="checkbox"
          value="GB"
        />
        <label
          for="GB"
        >
          United kingdom (GB)
        </label>
      </li>
    </ul>
  </main>
</div>
`;
